### PR TITLE
[trainer] solve "scheduler before optimizer step" warning

### LINKED ISSUE #11106: lr scheduler before optimizer step warning
for a really long time now we have this warning being spewed up on startup:

> torch/optim/lr_scheduler.py:132: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
>  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. 

Not sure if it's all the time, but sure thing a lot and it'd be nice to solve it.

I patched lr_scheduler's `step` to print the counters:
``` # torch/optim/lr_scheduler.py:117 
    def step(self, epoch=None):
        print(f"self._step_count={self._step_count}, self.optimizer._step_count={self.optimizer._step_count}")
```

With standard way we get:

```
net = torch.nn.Linear(1,1)
optimizer = torch.optim.SGD(net.parameters(), lr=1)
scheduler = torch.optim.lr_scheduler.LambdaLR(optimizer, lr_lambda=[lambda epoch: epoch // 30])

optimizer.step()
scheduler.step()

optimizer.step()
scheduler.step()

optimizer.step()
scheduler.step()
```
output:
```
self._step_count=0, self.optimizer._step_count=0
self._step_count=1, self.optimizer._step_count=1
self._step_count=2, self.optimizer._step_count=2
self._step_count=3, self.optimizer._step_count=3
```

But when we run some example with the HF Trainer, e.g. with:
```
export BS=1; rm -rf /tmp/test-clm; PYTHONPATH=src USE_TF=0 CUDA_VISIBLE_DEVICES=0 python \
examples/language-modeling/run_clm.py --model_name_or_path distilgpt2 --dataset_name wikitext \
--dataset_config_name wikitext-2-raw-v1 --do_train --max_train_samples 1 \
--per_device_train_batch_size $BS --output_dir /tmp/test-clm --block_size 128 --logging_steps 1 \
--fp16 --block_size=128 
```

we get:

```
self._step_count=0, self.optimizer._step_count=0
self._step_count=1, self.optimizer._step_count=0
self._step_count=2, self.optimizer._step_count=0
self._step_count=3, self.optimizer._step_count=1
```

as you can see the counter is off by 2 - that's why we get the warning. 

The order of the `step()` calls in the code looks ok, I think our optimizers may get reset after the scheduler was created perhaps?

@sgugger

### PR DESCRIPTION
**Pull request by stas00**
As discussed in https://github.com/huggingface/transformers/issues/11106 fp16 scaler leads to a warning:

> torch/optim/lr_scheduler.py:132: UserWarning: Detected call of lr_scheduler.step() before optimizer.step(). In PyTorch 1.1.0 and later, you should call them in the opposite order: optimizer.step() before lr_scheduler.step(). Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
> warnings.warn("Detected call of lr_scheduler.step() before optimizer.step().

because the optimizer may get skipped until the right scale is found, so we shouldn't run `lr_scheduler.step()` when that happens.

@ptrblck provided a workaround here:
https://discuss.pytorch.org/t/model-weights-not-getting-updated-when-using-autocast/117286/10?u=ptrblck

This is also reported at pytoch: https://github.com/pytorch/pytorch/issues/55585

So the solution is we check the scale before and after and if it changed, then the optimizer wasn't run and we skip the scheduler step then.

Fixes: https://github.com/huggingface/transformers/issues/11106

@sgugger

### CODE REVIEW

**Comment by sgugger** [comment]
I'm a bit torn about this solution: it solves the problem exposed by the warning but it creates another problem with no warning, which I had flagged in #11106 as 

> If we somehow manage to catch those skipped optimizer steps and delay the scheduler steps, then we won't respect the number of steps in the scheduler, leading to some wrong end learning rates.

I have no idea if it's better to skip the beginning or end values for the learning rate though.

Also the test is wrong I think it should be
```
optimizer_was_run = scale_before <= scale_after
```
as the scale factor can be multiplied by the `growth_factor` (after a long period without decrease) without skipping a step. It's when it decreases that we know the step was skipped.

**Comment by stas00** [comment]
> I'm a bit torn about this solution: it solves the problem exposed by the warning but it creates another problem with no warning, which I had flagged in #11106 as
> 
> > If we somehow manage to catch those skipped optimizer steps and delay the scheduler steps, then we won't respect the number of steps in the scheduler, leading to some wrong end learning rates.

If the scheduler and the optimizer are now synchronized why would this happen?

> ```
> optimizer_was_run = scale_before <= scale_after
> ```

Fixed, thank you for catching this!

**Comment by sgugger** [comment]
> If the scheduler and the optimizer are now synchronized why would this happen?

The scheduler was built with a certain number of total steps (for instance go linearly from 1e-4 to 0 in 500 steps). So by skipping those initial steps, we won't be seeing the last learning rates.

**Comment by stas00** [comment]
> The scheduler was built with a certain number of total steps (for instance go linearly from 1e-4 to 0 in 500 steps). So by skipping those initial steps, we won't be seeing the last learning rates.

Yes! My apologies I understood this from your earlier comment. I just meant that until this PR the scheduler was not synced with optimizer.

So best to truncate the last learning rates where it's usually fixed, or doesn't quite matter if it's a huge run and if it is cyclical it doesn't matter for sure if I understand the situation correctly. Let's perhaps ask it differently - in what situations do you think this mismatch/missing few last steps would practically matter?

**Comment by sgugger** [comment]
Thinking more and I think it's actually better to skip at the end, since the "short" part of a scheduler is often the warmup (for instance if we set `warum_steps=50`). So I revert my previous objection and I'm okay with the PR :-)

**Review by sgugger — APPROVED** [review]

**Comment by stas00** [comment]
That's a very good point!

And this will also synchronize with the behavior one gets under deepspeed.

Thank you for this brainstorming, @sgugger!
